src/dftbp/dftb/multiexpan.F90

```diff
@@ -0,0 +1,1742 @@
+    @:ASSERT(size(M3x3x3, dim=1) == dimSize)
+    @:ASSERT(size(M3x3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3, dim=3) == dimSize)
+    do ii = 1, dimSize
+      do jj = 1, dimSize
+        do ll = 1, dimSize
+          M3x3x3(ii,jj,ll) = A3x3(ii,ll) * B3(jj)
+        end do
+      end do
+    end do
+
+  end subroutine makeA3x3otB3
+
+
+  subroutine makeA3x3oB3x3(M3x3x3x3, A3x3, B3x3)
+
+    real(dp), intent(out) :: M3x3x3x3(:,:,:,:)
+    real(dp), intent(in) :: A3x3(:,:), B3x3(:,:)
+    integer :: ii, jj, ll, mm, dimSize
+
+    dimSize=3
+    @:ASSERT(size(B3x3, dim=1) == dimSize)
+    @:ASSERT(size(B3x3, dim=2) == dimSize)
+    @:ASSERT(size(A3x3, dim=1) == dimSize)
+    @:ASSERT(size(A3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=1) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=3) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=4) == dimSize)
+    do ii = 1, dimSize
+      do jj = 1, dimSize
+        do ll = 1, dimSize
+          do mm = 1, dimSize
+            M3x3x3x3(ii,jj,ll,mm) = A3x3(ii,jj) * B3x3(ll,mm)
+          end do
+        end do
+      end do
+    end do
+
+  end subroutine makeA3x3oB3x3
+
+
+  subroutine makeA3x3ohB3x3(M3x3x3x3, A3x3, B3x3)
+
+    real(dp), intent(out) :: M3x3x3x3(:,:,:,:)
+    real(dp), intent(in) :: A3x3(:,:), B3x3(:,:)
+    integer :: ii, jj, ll, mm, dimSize
+
+    dimSize=3
+    @:ASSERT(size(B3x3, dim=1) == dimSize)
+    @:ASSERT(size(B3x3, dim=2) == dimSize)
+    @:ASSERT(size(A3x3, dim=1) == dimSize)
+    @:ASSERT(size(A3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=1) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=3) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=4) == dimSize)
+    do ii = 1, dimSize
+      do jj = 1, dimSize
+        do ll = 1, dimSize
+          do mm = 1, dimSize
+            M3x3x3x3(ii,jj,ll,mm) = A3x3(ii,ll) * B3x3(jj,mm)
+          end do
+        end do
+      end do
+    end do
+
+  end subroutine makeA3x3ohB3x3
+
+
+  subroutine makeA3x3otB3x3(M3x3x3x3, A3x3, B3x3)
+
+    real(dp), intent(out) :: M3x3x3x3(:,:,:,:)
+    real(dp), intent(in) :: A3x3(:,:), B3x3(:,:)
+    integer :: ii, jj, ll, mm, dimSize
+
+    dimSize=3
+    @:ASSERT(size(B3x3, dim=1) == dimSize)
+    @:ASSERT(size(B3x3, dim=2) == dimSize)
+    @:ASSERT(size(A3x3, dim=1) == dimSize)
+    @:ASSERT(size(A3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=1) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=2) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=3) == dimSize)
+    @:ASSERT(size(M3x3x3x3, dim=4) == dimSize)
+    do ii = 1, dimSize
+      do jj = 1, dimSize
+        do ll = 1, dimSize
+          do mm = 1, dimSize
+            M3x3x3x3(ii,jj,ll,mm) = A3x3(ii,mm) * B3x3(jj,ll)
+          end do
+        end do
+      end do
+    end do
+
+  end subroutine makeA3x3otB3x3
+
+  !> copy lower triangle to upper for a square matrix
+  subroutine symmetrizeSquareMatrix(matrix)
```

Comment:
Duplicates functionality of the (actually imported above, but not used) adjointLowerTriangle routine. Also, called for a 3x3 matrix, so probably shouldn't be OMP parallel.